### PR TITLE
Objstore: Update objstore to latest version

### PR DESCRIPTION
**Pull request by phillebaba**
* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

Updates the objstore dependency to the latest version. Additionally the docs have been updated to reflect the requirement of go 1.18 or higher when developing Thanos. This is required because the new Azure SDK dependency introduced in objstore requires go 1.18 due to its use of generics.

## Verification

Built a fork from 0.28.0 with the updated objstore version and deployed it, has been running without any issues.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -24,6 +24,7 @@ We use *breaking :warning:* to mark changes that are not backward compatible (re
 
 - [#5255](https://github.com/thanos-io/thanos/pull/5296) Query: Use k-way merging for the proxying logic. The proxying sub-system now uses much less resources (~25-80% less CPU usage, ~30-50% less RAM usage according to our benchmarks). Reduces query duration by a few percent on queries with lots of series.
 - [#5690](https://github.com/thanos-io/thanos/pull/5690) Compact: update `--debug.accept-malformed-index` flag to apply to downsampling. Previously the flag only applied to compaction, and fatal errors would still occur when downsampling was attempted.
+- [#5707](https://github.com/thanos-io/thanos/pull/5707) Objstore: Update objstore to latest version.
```
Comment (fpetkovski) [inline CHANGELOG.md:27]: Is there something specific which the latest version contains that we can mention here?

Comment (phillebaba) [reply CHANGELOG.md:27]: Yes the main reason for updating is that Thanos will use the refactored Azure implementation.

File: .github/workflows/codeql-analysis.yml [hunk 2]
```diff
@@ -37,6 +37,11 @@ jobs:
     - name: Checkout repository
       uses: actions/checkout@v2
 
+    - name: Set up Go
+      uses: actions/setup-go@v2
+      with:
+        go-version: 1.18
```
Comment (phillebaba) [inline .github/workflows/codeql-analysis.yml:43]: Same change was made in objstore. Required for codeql to pass with libraries using generics.

**Review by fpetkovski — DISMISSED** [review]
Thanks 👍

**Review by bwplotka — DISMISSED** [review]
LGTM, thank You!

**Comment by bwplotka** [comment]
Looks like we have some important error to fix: https://app.circleci.com/pipelines/github/thanos-io/thanos/10720/workflows/81ea74ef-96bc-4d53-933b-4ad9c86d9080/jobs/20317

**Comment by phillebaba** [comment]
Looks like issues related to not skipping new objstore implementations. Will fix this.

**Comment by phillebaba** [comment]
@bwplotka @fpetkovski a bit confused about the failing e2e tests. The only changes done to objstore is the introduction of OCI storage and my Azure refactor. None of these things are testsed in the e2e test, so dont really understand how it is failing now.

**Comment by phillebaba** [comment]
Looks like the e2e test failure was due to flakiness in the test. I have rebased so should be ready to merge.

**Review by fpetkovski — APPROVED** [review]
Thanks for fixing the CI 👍

**Review by squat — APPROVED** [review]
Thanks @phillebaba

**Pull request merged**